### ISSUE TITLE
wiki: Skip Redirect : ELI5

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#249]
* [Skip Redirect](https://addons.mozilla.org/firefox/addon/skip-redirect/) | [GitHub](https://github.com/sblask/webextension-skip-redirect)

@rekixex suggested that we add Skip Redirect to the wiki. I am starting a new topic to cover this, as I would like someone to explain why we need another extension

I have to admit I am a little confused - I still have Redirector disabled (too f**k hard for me to write rules) and Request Control has almost nothing in it (I still have trouble writing rules!). I have NoRedirect (legacy) about to go bye bye in 57Land (I used this to block ALL redirects by default) and I want to replace that. Now there is this Skip Redirect, and I am totally confused as to what they all achieve as a mix. Part of the problem is the names - three of these all use the word "redirect" in there titles which is confusing/misleading.

I want someone to ELI5 how and why Skip Redirect differs. Also, from those screenshots on AMO it looks like it annoys all the time with doorhanger notifications

**Comment by crssi** [comment]
Didn't I already suggest this one? Maybe forgot since I was sure of it.
Why it is different? Because it works. lol
No seriously, the skip redirects works, with minor tweaks, without false positives.
The most important part was when the author implemented the same domain detection to avoid skipping on those, which eliminated almost every non-desired skipping redirect, specially where the action would break login. See [here](https://github.com/sblask/webextension-skip-redirect/issues/30#issuecomment-314689662)

Redirector is not match to that for skipping purposes and it has other problems.

Its almost set and forget.
I have added those additional to default blacklist in the options:
```
cred_submit
/_login
/adfs
rover.ebay.co
```

**Comment by rekixex** [comment]
@Thorin-Oakenpants 

> from those screenshots on AMO it looks like it annoys all the time with doorhanger notifications

Notifications can be disabled - uncheck 'Enable popup' in the add-on's preferences and it will work _silently_.

@crssi 

> Why it is different? Because it works.

> Its almost set and forget.

That's why it should be considered.

What about the _security_ of this add-on? How does it work exactly?

**Comment by grauenwolfe** [comment]
Been using Skip Redirect (along with Link Cleaner) for a few months now and it works well enough by default and has been a good replacement for NoRedirect and Redirect Remover. SR has simple, basic preferences like blacklist, whitelist, etc. So far, so good, no complaints from me. I'll include a screenshot even though there's probably already one on it's AMO page.

![skip redirect](https://user-images.githubusercontent.com/25858089/31040360-2665e450-a54c-11e7-967c-a285d96eeb71.png)

**Comment by Thorin-Oakenpants** [comment]
> Didn't I already suggest this one?

It's been in the wiki sticky issue pending list for ages. This is about adding it to the actual wiki. I do not want the wiki to end up with hundreds of extensions with duplicity and overlap (and we already have some of that) - I just want the best of the best

### Duplicity?
- Neat URL + Link Cleaner : Neat URL can do everything Link Cleaner does and you can edit the parameters, so maybe Link Cleaner can get removed
- Redirector and Request Control - both "look" like they do the same thing

### ELI5
Here are my NoRedirect rules. They are used in order from bottom to top. The bottom rule is a BLOCK ALL, and then for me to be able to double-click RSS feed items in "Quite RSS" (standalone software program) I had to allow a few destinations as they use things like `http://feedproxy.google.com/~r/SITE/~STRING/`. And a couple of other sites for them to work (eg when I download PAFs from portablepps.com => sourceforge) etc. So default deny all and allow a tiny handful - set and forget forever!.

![noredirect](https://user-images.githubusercontent.com/16656956/31041751-07c73606-a5f7-11e7-81f4-1fdb0efc11b6.png)

So looking at grauenwolfe's piccy, this is the same right? I can block all in the `Mode` options, and create a whitelist.

So this is a great replacement for NoRedirect - right? In other words, all that link shortening crap can die, right? In my world, if people can't give me a direct link, so I know where I'm going, then screw them!

**Comment by earthlng** [comment]
NoRedirect *blocks* redirects while this and any other such addons *skip* redirects! Totally different thing

**Comment by Thorin-Oakenpants** [comment]
So what use is skip redirect then? I don't get it

PS: Thank F someone ELI5 as requested

**Comment by crssi** [comment]
^^ It goes directly to landing page, without some "middle" tracking URL.
I hope I did write in understanding manner.

**Comment by Atavic** [comment]
[redirect-control](https://github.com/hjgwmvya/redirect-control) is functional, allowing to see the urls and accept or refuse the redirection.

[/* 2619: limit HTTP redirects](https://github.com/ghacksuserjs/ghacks-user.js/blob/908a8d66f4987568fc76649a7f47e8f61f8588f2/user.js#L1292)

**Comment by grauenwolfe** [comment]
> NoRedirect blocks redirects while this and any other such addons skip redirects! Totally different thing

As you're own understanding of blocking vs. skipping, please explain what you're saying here? Not challenging here actually genuinely curious. What would be different outcomes of the two in relation to the desired end result. You may well be on to something but only seems a syntactical difference.

Some redirecting extensions make connections to the redirected site before sending you to the intended one, as @crssi  has already mentioned. From all I've been able to gather, Skip Redirect does not do this. I definitely could be wrong on this but this is what made me go with SR over others being offered. Seriously, I need to know if SR is not doing what I'm expecting of it.

**Comment by Thorin-Oakenpants** [comment]
@grauenwolfe I no longer know dude .. I'm like all confused :headspin: ... 

I block all "redirects" in NoRedirect - this means things like an rss google feedly url will not redirect. In 3 or so years since I installed NoRedirect, I have only ever had to add those few items in the pic up above. Haven't changed it in years. I have also blocked all that utm rubbish - not sure if that counts as a redirect. Maybe NoRedirect only works with a few parameters like that `~r/SITE`. I have no idea. I think the word redirect used in these extension names causes confusion

So IMO, I cannot understand why anyone want to be redirected. By this I mean the link/url lands on one page and ends up on another. I totally get the same domain issue wanting to be excluded from blocks.

What annoys me about this kind of approach in Skip Redirect, is it doesn't block anything if I understand it correctly. You can only "skip" either a whitelist or blacklist. Those not being "skipped" still end up on the eventual page, i.e nothing is done with them. And those being "skipped" do what - attempt to skip to the end and you still end up there. What purpose does this serve. I just don't get it. Is it skipping tracking pages in the middle or something?

tld;dr: WTF is the **blocking** redirects extension

**Comment by rekixex** [comment]
Can someone explain, does **Skip Redirect** _really_ improve privacy (like **Neat URL** does, by removing tracking parameters from URLs)? If I understand correctly, it prevents Firefox from visiting the original URL.

From **Link Cleaner** description:

> **Comparison to other addons**
> Unlike a plugin like the (outdated) CleanLinks, Link Cleaner doesn't inject
> JavaScript into pages to change links.
> Instead, it listens to main url requests and changes them to remove redirects
> or tracking.

Do **Skip Redirect** and **Neat URL** work similarly to **Link Cleaner**, without _possible negative security consequences_?

**Comment by grauenwolfe** [comment]
I decided to just ask sblask directly.
[https://github.com/sblask/webextension-skip-redirect/issues/50](url)

We'll see what he says, if anything. Definitely one of the less chatty devs out there.

**Comment by siric1** [comment]
> And those being "skipped" do what - attempt to skip to the end and you still end up there. What purpose does this serve. I just don't get it. Is it skipping tracking pages in the middle or something?

Example:

https://out.reddit.com/t3_73jhqt?url=http%3A%2F%2Fwww.businessinsider.com%2Fwhy-political-ads-should-be-regulated-online-2017-9%3F&token=AQAAbrPQWTqwpYRVaStC0maxIdRe-WaW3aFH_uZyn8SQmX1H87pX&app_name=reddit.com

If you visit this "redirect" URL, Reddit will know that you have visited the article on the businessinsider website. The Skip Redirect extension _extracts_ the businessinsider URL from the redirect URL and goes there directly, without ever visiting out.reddit.com. This prevents Reddit from knowing that you have visited that specific URL. Useful in cases where you simply want to visit a link but don't want to let the parent site know which URL you are visiting.

Then there are extensions that flat out _block_ the redirect so you don't end up at the end URL. Useful if you don't know what the end URL is going to be (could be malicious) - ie. through URL shortener services, which "obfuscate" the end URL.

The Neat URL extension serves a completely different purpose: it strips _tracking parameters_ from URLs - such as those set by Google Analytics, ie. utm_*.

**Comment by earthlng** [comment]
> > NoRedirect blocks redirects while this and any other such addons skip redirects! Totally different thing
>
> As you're own understanding of blocking vs. skipping, please explain what you're saying here? 

@siric1's comment explains it perfectly. Thanks @siric1

**Comment by grauenwolfe** [comment]
Thanks to both you. 

> Then there are extensions that flat out block the redirect so you don't end up at the end URL. Useful if you don't know what the end URL is going to be (could be malicious) - ie. through URL shortener services, which "obfuscate" the end URL.

OK, essentially just blocking malicious sites. Yes, I understand perfectly well the concept of "blocking" in that regard but to me seems independent of cleaning links. 

> The Neat URL extension serves a completely different purpose: it strips tracking parameters from URLs - such as those set by Google Analytics, ie. utm_*.

With the exception of "utm" not being included by default, how is this not exactly what Skip Redirect is doing? Sorry, if I'm misunderstanding here or are we saying the same thing?

**Comment by acydwarp** [comment]
> OK, essentially just blocking malicious sites. Yes, I understand perfectly well the concept of "blocking" in that regard but to me seems independent of cleaning links.
Redirect blockers such as NoRedirect block any type of redirect. So for example, https://bit.ly/1bNsLrK simply will stop loading. Since we don't know what's behind that bit.ly link (it could for example lead us to a malicious site, ip logger etc.), the block would protect against this. So yes, that is completely independent of of cleaning links.

> With the exception of "utm" not being included by default, how is this not exactly what Skip Redirect is doing? Sorry, if I'm misunderstanding here or are we saying the same thing?
Neat URL does include the utm parameter by default, it uses the following default param list (and also provides a way to add more params to this default list through it's settings): https://github.com/Smile4ever/firefoxaddons/blob/master/Neat%20URL-webext/options.js#L4

Skip Redirect does _not_ remove tracking params from the URL, it simply skips the redirects as I explained 2 posts earlier, by extracting the end URL from the redirect URL and thus preventing the parent site from knowing which site you are visiting.

**Comment by siric1** [comment]
> OK, essentially just blocking malicious sites. Yes, I understand perfectly well the concept of "blocking" in that regard but to me seems independent of cleaning links.

Redirect blockers such as NoRedirect block any type of redirect. So for example, https://bit.ly/1bNsLrK simply will stop loading. Since we don't know what's behind that bit.ly link (it could for example lead us to a malicious site, ip logger etc.), the block would protect against this. So yes, this is completely independent of of cleaning links.

> With the exception of "utm" not being included by default, how is this not exactly what Skip Redirect is doing? Sorry, if I'm misunderstanding here or are we saying the same thing?

Neat URL does include the utm parameter by default, it uses the following default param list (and also provides a way to add more params to this default list through it's settings): https://github.com/Smile4ever/firefoxaddons/blob/master/Neat%20URL-webext/options.js#L4

Skip Redirect does _not_ remove tracking params from the URL, it simply skips the redirects as I explained 2 posts earlier, by extracting the end URL from the redirect URL, visiting it directly and thus preventing the parent site from knowing which site you are visiting.

**Comment by grauenwolfe** [comment]
> Skip Redirect does not remove tracking params from the URL, it simply skips the redirects as I explained 2 posts earlier...

I know what you said two posts earlier. I have been mistaken about Skip Redirect and was under the impression that this is what it has been doing all along.

What still isn't clear is how will it matter in the end? Not rhetorically, I'm genuinely asking. Give a worst-case scenario example, even if absurd in order to make the point clear. Baby steps. If you haven't the patience, I'll understand.

**Comment by Thorin-Oakenpants** [comment]
- added to wiki I've installed it myself at defaults except i turned of notifcations
- sblask replied to grauenwolfe's Q: "Are any connections made to unfiltered links before sending to filtered links"
    * > As I use a blocking onBeforeRequest (https://github.com/sblask/webextension-skip-redirect/blob/master/background.js#L187) no connection should be made unless the browser does some funny stuff. ... Even if it seems so because the unfiltered URL appears in the location bar until the final page is loaded...
- This is all a bit muddled to me (and I'm not the only one - not all of us are internet rock star tech devs), and it doesn't help with the extension names. There are so many ways for sites to know how you got there (tracking parameters, redirects) and it all gets a bit confusing. Thanks all and @siric1 for your detailed explanations.
- Where the F is my NoRedirect web extension equivalent? I reckon a rockstar internet tech dev guru could bang that out in a day

**Comment by rekixex** [comment]
@Thorin-Oakenpants Now when Neat URL and Skip Redirect are in the wiki, is **Link Cleaner** redundant?

**Comment by Thorin-Oakenpants** [comment]
Link Cleaner strip parameters like utm etc and is a set list. And to quote siric1
> Neat URL extension serves a completely different purpose: it strips tracking parameters from URLs - such as those set by Google Analytics, ie. utm_*.

So Link Cleaner & Neat URL are pretty much the same, and as per siric1's comment, Skip Redirect is completely different. See my question about 5 comments in, and I asked if Link Cleaner & Neat URL were duplicitous.

**Comment by rekixex** [comment]
> Link Cleaner & Neat URL are pretty much the same, and as per siric1's comment, Skip Redirect is completely different.

Actually, not _completely_ different - Link Cleaner skips redirects on Facebook and Reddit. It removes utm_* parameters - Neat URL does the same, at least for some of them, but also can be configured to remove any known tracking parameter.

What's the benefit to use Link Cleaner alongside Neat URL and Skip Redirect? Just curious, I had all of them enabled without visible negative effects.

**Comment by Thorin-Oakenpants** [comment]
I have no idea .. its why I asked somebody to ELI5

**Comment by crssi** [comment]
Link Cleaner tries to do both as Skip Redirect and Neat URL does, but it lacks custom settings and user control. Don't get me wrong here, I am not saying that LC is doing worse for what is hardcoded in, but it is not trough as SR and NU.
SR does skipping redirect better (user control) and being "silent" when the "main" URL and redirection URL domain is the same, thus avoiding "false positives" and skipping redirect is in those cases a nonsense and leads to breakages.
NU does parameters cleaning better in a way that and the cleaned parameters are known and user controllable with possibility to add/extend the operations.

When using SR and NU, then IMO LC is redundant. Using SR and NU over LC is better solution.

**Comment by siric1** [comment]
Link Cleaner: Tries to strip url parameters as well as skip redirects, but on a very _limited set_. The major downside is the set is _hard-coded_ and the extension provides no configuration to add to it's default list. It can only clean utm_* and some tracking params on amazon and ali express. And, it only skips redirects for facebook, steam and reddit: 
https://github.com/idlewan/link_cleaner/blob/master/background.js#L164
 
Skip Redirect: Skips redirects _globally_ on all sites. Uses some smart filters to do this instead of a hard-coded list.

Neat URL: Strips many URL parameters by default and has the option to add even more.

The Skip Redirect + Neat URL combination is the clear winner here. Global redirect skipping and an unlimited set of tracking parameters as opposed Link Cleaner's 3-4 tracking parameters and 3 redirect skips is a no-brainer to me.

> Where the F is my NoRedirect web extension equivalent? I reckon a rockstar internet tech dev guru could bang that out in a day

Did a quick search for you but there seems to indeed be no extension that can block redirects entirely. I might look into creating one if the extension API provides me with enough tools to do it.

**Comment by crssi** [comment]
^^ :)

**Comment by Thorin-Oakenpants** [comment]
I'm using NoRedirect right now. I got a wee popup from Skip Redirect on a reddit link earlier (I turned on notifications because I want to see if its working), so either NoRedirect has broken recently (I am on 56) or it doesn't cover what you think it does, or it's no longer comprehensive .. i have no idea. All I know is I used this for years with no issues (and it worked initially because I had to add some exceptions) ... maybe its been broken for years, but earthlng uses this and he lives in his console ... and no it can't  be broken because I sometimes click shortened links for fun to see what happens = answer nothing = blocked

**Comment by Thorin-Oakenpants** [comment]
If LC is redundant, then if someone can confirm that all the tracking parameters in LC are in NU, then I can remove it

**Comment by crssi** [comment]
From [source code](https://github.com/idlewan/link_cleaner/blob/master/background.js) and from [manifest](https://github.com/idlewan/link_cleaner/blob/master/manifest.yaml)
LC cleans only:
  - removes utm_* parameters
  - on item pages of aliexpress and amazon, removes tracking parameters
  - skip redirect pages of facebook, steam and reddit (directly go to the url being redirected to, and never hit their outgoing redirect tracking)

So LC is very very limited.

**Comment by earthlng** [comment]
> I got a wee popup from Skip Redirect on a reddit link earlier so either NoRedirect has broken recently

With a skipping addon you never load out.reddit.com and therefore NoRedirect never kicks in.

NoRedirect is useful if for example the linked-to URL is another redirect url (bitly, etc) that could take you to a malicious site.

**Comment by Thorin-Oakenpants** [comment]
^^ OK that makes sense because NoRedirect only cares if you change domain - damn I hope that becomes a WE one day

**Comment by Thorin-Oakenpants** [comment]
So VOTE then .. remove LC if it's redundant and limited

**Comment by earthlng** [comment]
> NoRedirect only cares if you change domain

no it does not! move your blocking rule to the top and then go to https://developer.mozilla.org/en-US/Add-ons/WebExtensions and click on the "alarms" link under References for example

**Comment by Thorin-Oakenpants** [comment]
This is why I hate talking about redirects - i cannot wrap my head around all this shit

**Comment by Thorin-Oakenpants** [comment]
^^ done
